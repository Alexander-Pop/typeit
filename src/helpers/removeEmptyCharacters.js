import removeNode from "./removeNode";
import nodeCollectionToArray from "./nodeCollectionToArray";

/**
 * Check if a given character node is empty, either containing nothing or an empty HTML element.
 *
 * @param {object} node
 * @return {boolean}
 */
export const characterIsEmpty = node => {
  let childrenNodes = nodeCollectionToArray(node.childNodes);

  // There's nothing in here at all. Empty!
  if (!childrenNodes.length) {
    return true;
  }

  let childrenNode = childrenNodes[0];

  // Special treatment for <br> tags. These do not count as "empty."
  if (childrenNode.tagName === "BR") {
    return false;
  }

  // This is a text node. It's technically not empty.
  if (childrenNode.nodeType === 3) {
    return false;
  }

  // This is an empty HTML element!
  return !nodeCollectionToArray(childrenNodes[0].childNodes).length;
};

/**
 * Check if array of nodes contains any empty characters.
 *
 * @param {array} nodes
 * @return {boolean}
 */
export const containsEmptyCharacter = nodes => {
  return nodes.some(node => {
    return characterIsEmpty(node);
  });
};

/**
 * Given a DOM scope and selector, remove the empty `.ti-chars` nodes.
 *
 * @param {object} scope
 * @param {string} selector
 * @return {void}
 */
export default (scope, selector) => {
  let allChars = nodeCollectionToArray(scope.querySelectorAll(selector));
  let hasEmptyNodes = containsEmptyCharacter(allChars);

  while (allChars.length && hasEmptyNodes) {
    allChars.forEach(char => {
      if (characterIsEmpty(char)) {
        removeNode(char);
      }
    });

    allChars = nodeCollectionToArray(scope.querySelectorAll(selector));

    // Removing nodes might have created new empty nodes,
    // so we need to re-query and check again.
    hasEmptyNodes = containsEmptyCharacter(allChars);
  }

  return allChars;
};
