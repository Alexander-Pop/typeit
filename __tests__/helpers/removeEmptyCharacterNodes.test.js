import nodeCollectionToArray from "../../src/helpers/nodeCollectionToArray";
import removeEmptyCharacters, {
  characterIsEmpty,
  containsEmptyCharacter
} from "../../src/helpers/removeEmptyCharacters";

describe("characterIsEmpty()", () => {
  test("Returns false when a break tag is encountered.", () => {
    document.body.innerHTML = `<i class="ti-char"><br></i>`;
    let node = document.querySelector(".ti-char");
    let result = characterIsEmpty(node);
    expect(result).toBe(false);
  });

  test("Returns true when character is empty.", () => {
    document.body.innerHTML = `<i class="ti-char"></i>`;
    let node = document.querySelector(".ti-char");
    let result = characterIsEmpty(node);
    expect(result).toBe(true);
  });

  test("Returns false when character has text.", () => {
    document.body.innerHTML = `<i class="ti-char">hi.</i>`;
    let node = document.querySelector(".ti-char");
    let result = characterIsEmpty(node);
    expect(result).toBe(false);
  });

  test("Returns true when character has empty HTML", () => {
    document.body.innerHTML = `<i class="ti-char"><span></span></i>`;
    let node = document.querySelector(".ti-char");
    let result = characterIsEmpty(node);
    expect(result).toBe(true);
  });

  test("Returns false when character contains HTML containing other character (empty or not).", () => {
    document.body.innerHTML = `<i class="ti-char"><span><i class="ti-char"><a></a></i></span></i>`;
    let node = document.querySelector(".ti-char");
    let result = characterIsEmpty(node);
    expect(result).toBe(false);
  });
});

describe("containsEmptyCharacter()", () => {
  test("Returns true when at least one character is empty.", () => {
    document.body.innerHTML = `<i class="ti-char"></i><i class="ti-char">hi.</i>`;
    let nodes = nodeCollectionToArray(document.querySelectorAll(".ti-char"));
    let result = containsEmptyCharacter(nodes);
    expect(result).toBe(true);
  });

  test("Returns false when none are empty.", () => {
    document.body.innerHTML = `<i class="ti-char">sup.</i><i class="ti-char">hi.</i>`;
    let nodes = nodeCollectionToArray(document.querySelectorAll(".ti-char"));
    let result = containsEmptyCharacter(nodes);
    expect(result).toBe(false);
  });
});

describe("removeEmptyCharacters()", () => {
  test("Empty character should be removed from DOM.", () => {
    document.body.innerHTML = `
      <span id="scope">
        <i class="ti-char"></i>
        <i class="ti-char">a</i>
        <i class="ti-char">b</i>
        <i class="ti-char">c</i>
      </span>`;

    removeEmptyCharacters(document.getElementById("scope"), ".ti-char");
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test("Empty HTML character should be removed from DOM.", () => {
    document.body.innerHTML = `
      <span id="scope">
        <i class="ti-char">a</i>
        <i class="ti-char"><span></span></i>
        <i class="ti-char">b</i>
        <i class="ti-char">c</i>
      </span>`;

    removeEmptyCharacters(document.getElementById("scope"), ".ti-char");
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  test("Nested empty HTML character should be removed from DOM.", () => {
    document.body.innerHTML = `
      <span id="scope">
        <i class="ti-char">a</i>
        <i class="ti-char">b</i>
        <i class="ti-char"><span><i class="ti-char"><a></a></i></span></i>
        <i class="ti-char">c</i>
      </span>`;

    removeEmptyCharacters(document.getElementById("scope"), ".ti-char");
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
