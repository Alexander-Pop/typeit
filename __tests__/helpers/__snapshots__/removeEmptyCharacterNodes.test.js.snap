// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeEmptyCharacters() Empty HTML character should be removed from DOM. 1`] = `
"
      <span id=\\"scope\\">
        <i class=\\"ti-char\\">a</i>
        
        <i class=\\"ti-char\\">b</i>
        <i class=\\"ti-char\\">c</i>
      </span>"
`;

exports[`removeEmptyCharacters() Empty character should be removed from DOM. 1`] = `
"
      <span id=\\"scope\\">
        
        <i class=\\"ti-char\\">a</i>
        <i class=\\"ti-char\\">b</i>
        <i class=\\"ti-char\\">c</i>
      </span>"
`;

exports[`removeEmptyCharacters() Nested empty HTML character should be removed from DOM. 1`] = `
"
      <span id=\\"scope\\">
        <i class=\\"ti-char\\">a</i>
        <i class=\\"ti-char\\">b</i>
        
        <i class=\\"ti-char\\">c</i>
      </span>"
`;
